

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="../../../">
<head>
  <meta charset="utf-8" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>snpio.popgenstats.genetic_distance &mdash; SNPio 1.2.1 documentation</title>
      <link rel="stylesheet" type="text/css" href="../../../_static/pygments.css?v=80d5e7a1" />
      <link rel="stylesheet" type="text/css" href="../../../_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="../../../_static/custom.css?v=859da57e" />

  
      <script src="../../../_static/jquery.js?v=5d32c60e"></script>
      <script src="../../../_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="../../../_static/documentation_options.js?v=ca842793"></script>
      <script src="../../../_static/doctools.js?v=9bcbadda"></script>
      <script src="../../../_static/sphinx_highlight.js?v=dc90522c"></script>
    <script src="../../../_static/js/theme.js"></script>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="../../../index.html" class="icon icon-home">
            SNPio
              <img src="../../../_static/snpio_logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="../../../search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul>
<li class="toctree-l1"><a class="reference internal" href="../../../getting_started.html">Getting Started</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../example_script.html">Example script</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../pop_gen_statistics.html">PopGenStatistics Documentation</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../changelog.html">Changelog</a></li>
<li class="toctree-l1"><a class="reference internal" href="../../../modules.html">snpio</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="../../../index.html">SNPio</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="../../../index.html" class="icon icon-home" aria-label="Home"></a></li>
          <li class="breadcrumb-item"><a href="../../index.html">Module code</a></li>
      <li class="breadcrumb-item active">snpio.popgenstats.genetic_distance</li>
      <li class="wy-breadcrumbs-aside">
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <h1>Source code for snpio.popgenstats.genetic_distance</h1><div class="highlight"><pre>
<span></span><span class="kn">import</span> <span class="nn">numpy</span> <span class="k">as</span> <span class="nn">np</span>
<span class="kn">import</span> <span class="nn">pandas</span> <span class="k">as</span> <span class="nn">pd</span>


<div class="viewcode-block" id="GeneticDistance">
<a class="viewcode-back" href="../../../snpio.popgenstats.html#snpio.popgenstats.genetic_distance.GeneticDistance">[docs]</a>
<span class="k">class</span> <span class="nc">GeneticDistance</span><span class="p">:</span>
<span class="w">    </span><span class="sd">&quot;&quot;&quot;Class for calculating Nei&#39;s genetic distance between populations.</span>

<span class="sd">    This class calculates genetic distance between populations using allele frequencies. The genetic distance is calculated using Nei&#39;s method, which is based on the proportion of shared alleles between populations. The genetic distance is calculated as ``-log(I)``, where I is the proportion of shared alleles between populations.</span>
<span class="sd">    &quot;&quot;&quot;</span>

    <span class="k">def</span> <span class="fm">__init__</span><span class="p">(</span><span class="bp">self</span><span class="p">,</span> <span class="n">genotype_data</span><span class="p">,</span> <span class="n">alignment_012</span><span class="p">,</span> <span class="n">logger</span><span class="p">,</span> <span class="n">plotter</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Initialize the GeneticDistance object.</span>

<span class="sd">        Args:</span>
<span class="sd">            genotype_data (GenotypeData): GenotypeData object containing genotype data.</span>
<span class="sd">            alignment_012 (np.ndarray): Genotype data in 012-encoded format.</span>
<span class="sd">            logger (logging.Logger): Logger object.</span>
<span class="sd">            plotter (snpio.plotting.Plotting): Plotting object.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span> <span class="o">=</span> <span class="n">genotype_data</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">alignment_012</span> <span class="o">=</span> <span class="n">alignment_012</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span> <span class="o">=</span> <span class="n">logger</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">plotter</span> <span class="o">=</span> <span class="n">plotter</span>

<div class="viewcode-block" id="GeneticDistance.calculate_neis_distance">
<a class="viewcode-back" href="../../../snpio.popgenstats.html#snpio.popgenstats.genetic_distance.GeneticDistance.calculate_neis_distance">[docs]</a>
    <span class="k">def</span> <span class="nf">calculate_neis_distance</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span>
        <span class="n">n_bootstraps</span><span class="p">:</span> <span class="nb">int</span> <span class="o">=</span> <span class="mi">0</span><span class="p">,</span>
        <span class="n">palette</span><span class="p">:</span> <span class="nb">str</span> <span class="o">=</span> <span class="s2">&quot;coolwarm&quot;</span><span class="p">,</span>
        <span class="n">supress_plot</span><span class="p">:</span> <span class="nb">bool</span> <span class="o">=</span> <span class="kc">False</span><span class="p">,</span>
    <span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate Nei&#39;s genetic distance between all pairs of populations.</span>

<span class="sd">        Optionally computes bootstrap-based p-values for each population pair if n_bootstraps &gt; 0.</span>

<span class="sd">        Nei&#39;s genetic distance is defined as D = -ln( Ī ), where Ī is the ratio of the average genetic identity to the geometric mean of the average homozygosities.</span>

<span class="sd">        Args:</span>
<span class="sd">            n_bootstraps (int): Number of bootstrap replicates to compute p-values. Defaults to 0 (only distances are returned).</span>
<span class="sd">            palette (str): Color palette for the distance matrix plot. Can use any matplotlib gradient-based palette. Some frequently used options include: &quot;coolwarm&quot;, &quot;viridis&quot;, &quot;magma&quot;, and &quot;inferno&quot;. Defaults to &#39;coolwarm&#39;.</span>
<span class="sd">            supress_plot (bool): If True, suppresses the plotting of the distance matrix. Defaults to False.</span>

<span class="sd">        Returns:</span>
<span class="sd">            pd.DataFrame: If n_bootstraps == 0, returns a DataFrame of Nei&#39;s distances.</span>
<span class="sd">            Tuple[pd.DataFrame, pd.DataFrame]: If n_bootstraps &gt; 0, returns a tuple of (distance matrix, p-value matrix).</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Calculating Nei&#39;s genetic distance...&quot;</span><span class="p">)</span>

        <span class="c1"># Step 1: Get allele frequencies per population.</span>
        <span class="n">allele_freqs_per_pop</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_allele_frequencies</span><span class="p">()</span>
        <span class="n">populations</span> <span class="o">=</span> <span class="nb">list</span><span class="p">(</span><span class="n">allele_freqs_per_pop</span><span class="o">.</span><span class="n">keys</span><span class="p">())</span>

        <span class="c1"># Initialize the distance matrix and, if needed, the p-value matrix.</span>
        <span class="n">dist_matrix</span> <span class="o">=</span> <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span>
            <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">populations</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">populations</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span>
        <span class="p">)</span>

        <span class="n">pval_matrix</span> <span class="o">=</span> <span class="p">(</span>
            <span class="n">pd</span><span class="o">.</span><span class="n">DataFrame</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">index</span><span class="o">=</span><span class="n">populations</span><span class="p">,</span> <span class="n">columns</span><span class="o">=</span><span class="n">populations</span><span class="p">,</span> <span class="n">dtype</span><span class="o">=</span><span class="nb">float</span><span class="p">)</span>
            <span class="k">if</span> <span class="n">n_bootstraps</span> <span class="o">&gt;</span> <span class="mi">0</span>
            <span class="k">else</span> <span class="kc">None</span>
        <span class="p">)</span>

        <span class="c1"># Step 2: Loop over all pairs.</span>
        <span class="k">for</span> <span class="n">i</span><span class="p">,</span> <span class="n">pop1</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">populations</span><span class="p">):</span>
            <span class="n">freqs_pop1</span> <span class="o">=</span> <span class="n">allele_freqs_per_pop</span><span class="p">[</span><span class="n">pop1</span><span class="p">]</span>
            <span class="k">for</span> <span class="n">j</span><span class="p">,</span> <span class="n">pop2</span> <span class="ow">in</span> <span class="nb">enumerate</span><span class="p">(</span><span class="n">populations</span><span class="p">):</span>
                <span class="k">if</span> <span class="n">i</span> <span class="o">==</span> <span class="n">j</span><span class="p">:</span>
                    <span class="n">dist_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop1</span><span class="p">,</span> <span class="n">pop2</span><span class="p">]</span> <span class="o">=</span> <span class="mf">0.0</span>
                    <span class="k">if</span> <span class="n">pval_matrix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">pval_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop1</span><span class="p">,</span> <span class="n">pop2</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
                <span class="k">elif</span> <span class="n">j</span> <span class="o">&lt;</span> <span class="n">i</span><span class="p">:</span>
                    <span class="c1"># Use symmetry: copy from the other half.</span>
                    <span class="n">dist_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop1</span><span class="p">,</span> <span class="n">pop2</span><span class="p">]</span> <span class="o">=</span> <span class="n">dist_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop2</span><span class="p">,</span> <span class="n">pop1</span><span class="p">]</span>
                    <span class="k">if</span> <span class="n">pval_matrix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
                        <span class="n">pval_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop1</span><span class="p">,</span> <span class="n">pop2</span><span class="p">]</span> <span class="o">=</span> <span class="n">pval_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop2</span><span class="p">,</span> <span class="n">pop1</span><span class="p">]</span>
                <span class="k">else</span><span class="p">:</span>
                    <span class="n">freqs_pop2</span> <span class="o">=</span> <span class="n">allele_freqs_per_pop</span><span class="p">[</span><span class="n">pop2</span><span class="p">]</span>
                    <span class="c1"># Calculate observed Nei&#39;s distance.</span>
                    <span class="n">D_obs</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">_calculate_neis_distance_between_pops</span><span class="p">(</span>
                        <span class="n">freqs_pop1</span><span class="p">,</span> <span class="n">freqs_pop2</span>
                    <span class="p">)</span>
                    <span class="n">dist_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop1</span><span class="p">,</span> <span class="n">pop2</span><span class="p">]</span> <span class="o">=</span> <span class="n">D_obs</span>
                    <span class="n">dist_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop2</span><span class="p">,</span> <span class="n">pop1</span><span class="p">]</span> <span class="o">=</span> <span class="n">D_obs</span>

                    <span class="c1"># If bootstrapping is requested, compute a p-value.</span>
                    <span class="k">if</span> <span class="n">n_bootstraps</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">:</span>
                        <span class="c1"># Identify loci with valid data.</span>
                        <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
                            <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">freqs_pop1</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">freqs_pop2</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isinf</span><span class="p">(</span><span class="n">freqs_pop1</span><span class="p">)</span>
                            <span class="o">&amp;</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isinf</span><span class="p">(</span><span class="n">freqs_pop2</span><span class="p">)</span>
                        <span class="p">)</span>
                        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">valid</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                            <span class="n">pval</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
                        <span class="k">else</span><span class="p">:</span>
                            <span class="n">p1_valid</span> <span class="o">=</span> <span class="n">freqs_pop1</span><span class="p">[</span><span class="n">valid</span><span class="p">]</span>
                            <span class="n">q_valid</span> <span class="o">=</span> <span class="n">freqs_pop2</span><span class="p">[</span><span class="n">valid</span><span class="p">]</span>
                            <span class="c1"># Clip allele frequencies to [0, 1] to avoid numerical issues.</span>
                            <span class="n">p1_valid</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">p1_valid</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                            <span class="n">q_valid</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">q_valid</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
                            <span class="n">L</span> <span class="o">=</span> <span class="nb">len</span><span class="p">(</span><span class="n">p1_valid</span><span class="p">)</span>
                            <span class="n">boot_dists</span> <span class="o">=</span> <span class="p">[]</span>
                            <span class="k">for</span> <span class="n">b</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="n">n_bootstraps</span><span class="p">):</span>
                                <span class="n">sample_idx</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">random</span><span class="o">.</span><span class="n">choice</span><span class="p">(</span><span class="n">L</span><span class="p">,</span> <span class="n">size</span><span class="o">=</span><span class="n">L</span><span class="p">,</span> <span class="n">replace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
                                <span class="n">p_boot</span> <span class="o">=</span> <span class="n">p1_valid</span><span class="p">[</span><span class="n">sample_idx</span><span class="p">]</span>
                                <span class="n">q_boot</span> <span class="o">=</span> <span class="n">q_valid</span><span class="p">[</span><span class="n">sample_idx</span><span class="p">]</span>

                                <span class="c1"># Compute bootstrap replicate of Nei&#39;s distance.</span>
                                <span class="n">I_locus</span> <span class="o">=</span> <span class="n">p_boot</span> <span class="o">*</span> <span class="n">q_boot</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p_boot</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">q_boot</span><span class="p">)</span>

                                <span class="n">I_bar</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">I_locus</span><span class="p">)</span>

                                <span class="n">homo1</span> <span class="o">=</span> <span class="n">p_boot</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p_boot</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span>
                                <span class="n">homo2</span> <span class="o">=</span> <span class="n">q_boot</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">q_boot</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span>

                                <span class="n">avg_homo1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">homo1</span><span class="p">)</span>
                                <span class="n">avg_homo2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">homo2</span><span class="p">)</span>

                                <span class="n">denom</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">avg_homo1</span> <span class="o">*</span> <span class="n">avg_homo2</span><span class="p">)</span>

                                <span class="k">if</span> <span class="n">denom</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
                                    <span class="n">boot_d</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
                                <span class="k">else</span><span class="p">:</span>
                                    <span class="n">I_ratio</span> <span class="o">=</span> <span class="n">I_bar</span> <span class="o">/</span> <span class="n">denom</span>
                                    <span class="k">if</span> <span class="n">I_ratio</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
                                        <span class="n">boot_d</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span>
                                    <span class="k">else</span><span class="p">:</span>
                                        <span class="n">boot_d</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">I_ratio</span><span class="p">)</span>
                                <span class="n">boot_dists</span><span class="o">.</span><span class="n">append</span><span class="p">(</span><span class="n">boot_d</span><span class="p">)</span>

                            <span class="n">boot_dists</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">array</span><span class="p">(</span><span class="n">boot_dists</span><span class="p">)</span>
                            <span class="n">valid_boot</span> <span class="o">=</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">boot_dists</span><span class="p">)</span>

                            <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">valid_boot</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
                                <span class="n">pval</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
                            <span class="k">else</span><span class="p">:</span>
                                <span class="c1"># One-tailed p-value: proportion of bootstrap replicates with distance &gt;= observed.</span>
                                <span class="n">pval</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">boot_dists</span><span class="p">[</span><span class="n">valid_boot</span><span class="p">]</span> <span class="o">&gt;=</span> <span class="n">D_obs</span><span class="p">)</span>

                        <span class="n">pval_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop1</span><span class="p">,</span> <span class="n">pop2</span><span class="p">]</span> <span class="o">=</span> <span class="n">pval</span>
                        <span class="n">pval_matrix</span><span class="o">.</span><span class="n">loc</span><span class="p">[</span><span class="n">pop2</span><span class="p">,</span> <span class="n">pop1</span><span class="p">]</span> <span class="o">=</span> <span class="n">pval</span>

        <span class="c1"># Replace any infinite values with NaN.</span>
        <span class="n">dist_matrix</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">pval_matrix</span> <span class="ow">is</span> <span class="ow">not</span> <span class="kc">None</span><span class="p">:</span>
            <span class="n">pval_matrix</span><span class="o">.</span><span class="n">replace</span><span class="p">(</span><span class="n">np</span><span class="o">.</span><span class="n">inf</span><span class="p">,</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span><span class="p">,</span> <span class="n">inplace</span><span class="o">=</span><span class="kc">True</span><span class="p">)</span>
            <span class="n">args</span> <span class="o">=</span> <span class="p">(</span><span class="n">dist_matrix</span><span class="p">,</span> <span class="n">pval_matrix</span><span class="p">)</span>
        <span class="k">else</span><span class="p">:</span>
            <span class="n">args</span> <span class="o">=</span> <span class="n">dist_matrix</span>

        <span class="k">if</span> <span class="ow">not</span> <span class="n">supress_plot</span><span class="p">:</span>
            <span class="bp">self</span><span class="o">.</span><span class="n">plotter</span><span class="o">.</span><span class="n">plot_dist_matrix</span><span class="p">(</span>
                <span class="o">*</span><span class="n">args</span><span class="p">,</span> <span class="n">palette</span><span class="o">=</span><span class="n">palette</span><span class="p">,</span> <span class="n">title</span><span class="o">=</span><span class="s2">&quot;Nei&#39;s Genetic Distance&quot;</span>
            <span class="p">)</span>

        <span class="bp">self</span><span class="o">.</span><span class="n">logger</span><span class="o">.</span><span class="n">info</span><span class="p">(</span><span class="s2">&quot;Nei&#39;s genetic distance calculation complete!&quot;</span><span class="p">)</span>

        <span class="k">return</span> <span class="n">args</span></div>


    <span class="k">def</span> <span class="nf">_calculate_neis_distance_between_pops</span><span class="p">(</span>
        <span class="bp">self</span><span class="p">,</span> <span class="n">freqs_pop1</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span><span class="p">,</span> <span class="n">freqs_pop2</span><span class="p">:</span> <span class="n">np</span><span class="o">.</span><span class="n">ndarray</span>
    <span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">float</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate Nei&#39;s genetic distance between two populations using averaged per-locus values.</span>

<span class="sd">        The function computes the per-locus genetic identity `(I = p*q + (1-p)*(1-q)) and then averages over loci. Nei&#39;s genetic distance is then defined as D = -ln( Ī ), where Ī is the ratio of the average identity to the geometric mean of the average homozygosities.</span>

<span class="sd">        Args:</span>
<span class="sd">            freqs_pop1 (np.ndarray): Allele frequencies for population 1.</span>
<span class="sd">            freqs_pop2 (np.ndarray): Allele frequencies for population 2.</span>

<span class="sd">        Returns:</span>
<span class="sd">            float: Nei&#39;s genetic distance. Returns np.nan if no valid loci exist, or np.inf if the</span>
<span class="sd">                identity ratio is non-positive.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="c1"># Identify loci with valid data.</span>
        <span class="n">valid</span> <span class="o">=</span> <span class="p">(</span>
            <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">freqs_pop1</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">freqs_pop2</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isinf</span><span class="p">(</span><span class="n">freqs_pop1</span><span class="p">)</span>
            <span class="o">&amp;</span> <span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isinf</span><span class="p">(</span><span class="n">freqs_pop2</span><span class="p">)</span>
        <span class="p">)</span>
        <span class="k">if</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="n">valid</span><span class="p">)</span> <span class="o">==</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

        <span class="n">p</span> <span class="o">=</span> <span class="n">freqs_pop1</span><span class="p">[</span><span class="n">valid</span><span class="p">]</span>
        <span class="n">q</span> <span class="o">=</span> <span class="n">freqs_pop2</span><span class="p">[</span><span class="n">valid</span><span class="p">]</span>
        <span class="c1"># Clip values to ensure they lie within [0, 1].</span>
        <span class="n">p</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">p</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>
        <span class="n">q</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">clip</span><span class="p">(</span><span class="n">q</span><span class="p">,</span> <span class="mi">0</span><span class="p">,</span> <span class="mi">1</span><span class="p">)</span>

        <span class="c1"># Compute per-locus genetic identity.</span>
        <span class="n">I_locus</span> <span class="o">=</span> <span class="n">p</span> <span class="o">*</span> <span class="n">q</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p</span><span class="p">)</span> <span class="o">*</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">q</span><span class="p">)</span>
        <span class="n">I_bar</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">I_locus</span><span class="p">)</span>

        <span class="c1"># Compute homozygosities.</span>
        <span class="n">homo1</span> <span class="o">=</span> <span class="n">p</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">p</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span>
        <span class="n">homo2</span> <span class="o">=</span> <span class="n">q</span><span class="o">**</span><span class="mi">2</span> <span class="o">+</span> <span class="p">(</span><span class="mi">1</span> <span class="o">-</span> <span class="n">q</span><span class="p">)</span> <span class="o">**</span> <span class="mi">2</span>
        <span class="n">avg_homo1</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">homo1</span><span class="p">)</span>
        <span class="n">avg_homo2</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">mean</span><span class="p">(</span><span class="n">homo2</span><span class="p">)</span>

        <span class="c1"># Compute the geometric mean of average homozygosities.</span>
        <span class="n">denom</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">sqrt</span><span class="p">(</span><span class="n">avg_homo1</span> <span class="o">*</span> <span class="n">avg_homo2</span><span class="p">)</span>
        <span class="k">if</span> <span class="n">denom</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>

        <span class="n">I_ratio</span> <span class="o">=</span> <span class="n">I_bar</span> <span class="o">/</span> <span class="n">denom</span>
        <span class="k">if</span> <span class="n">I_ratio</span> <span class="o">&lt;=</span> <span class="mi">0</span><span class="p">:</span>
            <span class="k">return</span> <span class="n">np</span><span class="o">.</span><span class="n">inf</span>

        <span class="n">nei_distance</span> <span class="o">=</span> <span class="o">-</span><span class="n">np</span><span class="o">.</span><span class="n">log</span><span class="p">(</span><span class="n">I_ratio</span><span class="p">)</span>
        <span class="k">return</span> <span class="n">nei_distance</span>

    <span class="k">def</span> <span class="nf">_calculate_allele_frequencies</span><span class="p">(</span><span class="bp">self</span><span class="p">)</span> <span class="o">-&gt;</span> <span class="nb">dict</span><span class="p">:</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;</span>
<span class="sd">        Helper method to calculate allele frequencies for each population at each locus.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: A dictionary where keys are population IDs and values are arrays of allele</span>
<span class="sd">            frequencies per locus.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pop_indices</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="o">.</span><span class="n">get_population_indices</span><span class="p">()</span>
        <span class="n">allele_freqs_per_pop</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">pop_id</span><span class="p">,</span> <span class="n">indices</span> <span class="ow">in</span> <span class="n">pop_indices</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="c1"># Subset alignment for population</span>
            <span class="n">pop_alignment</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignment_012</span><span class="p">[</span><span class="n">indices</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">pop_alignment</span><span class="p">[</span><span class="n">pop_alignment</span> <span class="o">==</span> <span class="o">-</span><span class="mi">9</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>  <span class="c1"># Replace missing data</span>

            <span class="c1"># Calculate allele frequencies</span>
            <span class="c1"># Since diploid</span>
            <span class="n">total_alleles</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">pop_alignment</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">alt_allele_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">pop_alignment</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>

            <span class="c1"># Avoid division by zero</span>
            <span class="n">freqs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">alt_allele_counts</span><span class="p">,</span> <span class="n">total_alleles</span><span class="p">,</span> <span class="n">where</span><span class="o">=</span><span class="n">total_alleles</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>

            <span class="n">allele_freqs_per_pop</span><span class="p">[</span><span class="n">pop_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">freqs</span>

        <span class="k">return</span> <span class="n">allele_freqs_per_pop</span>

    <span class="k">def</span> <span class="nf">_calculate_allele_frequencies</span><span class="p">(</span><span class="bp">self</span><span class="p">):</span>
<span class="w">        </span><span class="sd">&quot;&quot;&quot;Calculate allele frequencies for each population.</span>

<span class="sd">        This method calculates allele frequencies for each population in the genotype data. The allele frequencies are calculated as the proportion of alternate alleles at each locus.</span>

<span class="sd">        Returns:</span>
<span class="sd">            dict: Dictionary of allele frequencies for each population.</span>
<span class="sd">        &quot;&quot;&quot;</span>
        <span class="n">pop_indices</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">genotype_data</span><span class="o">.</span><span class="n">get_population_indices</span><span class="p">()</span>
        <span class="n">allele_freqs_per_pop</span> <span class="o">=</span> <span class="p">{}</span>

        <span class="k">for</span> <span class="n">pop_id</span><span class="p">,</span> <span class="n">indices</span> <span class="ow">in</span> <span class="n">pop_indices</span><span class="o">.</span><span class="n">items</span><span class="p">():</span>
            <span class="n">pop_alignment</span> <span class="o">=</span> <span class="bp">self</span><span class="o">.</span><span class="n">alignment_012</span><span class="p">[</span><span class="n">indices</span><span class="p">,</span> <span class="p">:]</span><span class="o">.</span><span class="n">astype</span><span class="p">(</span><span class="nb">float</span><span class="p">)</span><span class="o">.</span><span class="n">copy</span><span class="p">()</span>
            <span class="n">pop_alignment</span><span class="p">[</span><span class="n">pop_alignment</span> <span class="o">==</span> <span class="o">-</span><span class="mi">9</span><span class="p">]</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nan</span>
            <span class="n">total_alleles</span> <span class="o">=</span> <span class="mi">2</span> <span class="o">*</span> <span class="n">np</span><span class="o">.</span><span class="n">sum</span><span class="p">(</span><span class="o">~</span><span class="n">np</span><span class="o">.</span><span class="n">isnan</span><span class="p">(</span><span class="n">pop_alignment</span><span class="p">),</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">alt_allele_counts</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">nansum</span><span class="p">(</span><span class="n">pop_alignment</span><span class="p">,</span> <span class="n">axis</span><span class="o">=</span><span class="mi">0</span><span class="p">)</span>
            <span class="n">freqs</span> <span class="o">=</span> <span class="n">np</span><span class="o">.</span><span class="n">divide</span><span class="p">(</span><span class="n">alt_allele_counts</span><span class="p">,</span> <span class="n">total_alleles</span><span class="p">,</span> <span class="n">where</span><span class="o">=</span><span class="n">total_alleles</span> <span class="o">&gt;</span> <span class="mi">0</span><span class="p">)</span>
            <span class="n">allele_freqs_per_pop</span><span class="p">[</span><span class="n">pop_id</span><span class="p">]</span> <span class="o">=</span> <span class="n">freqs</span>

        <span class="k">return</span> <span class="n">allele_freqs_per_pop</span></div>

</pre></div>

           </div>
          </div>
          <footer>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2023, Bradley T. Martin and Tyler K. Chafin.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>